### PR TITLE
Handle in-use count errors

### PR DESCRIPTION
**Pull request by nafg**

### CODE REVIEW

**Comment by nafg** [comment]
Should resolve https://github.com/slick/slick/discussions/2432

**Comment by nafg** [comment]
@y-yu @dwickern please test

File: slick/src/main/scala/slick/util/AsyncExecutor.scala [hunk 1]
```diff
@@ -78,17 +97,68 @@ object AsyncExecutor extends Logging {
...
+            try queue.increaseInUseCount(pr)
+            catch {
+              case _: ManagedArrayBlockingQueue.CountLimitReachedException =>
+                logger.warn("Could not increase in-use count, because the limit has been reached. Will resubmit...")
+                executor.execute(r)
+            }
+          case _                                                        =>
+        }
+        super.beforeExecute(t, r)
+      }
```
Comment (dwickern) [inline slick/src/main/scala/slick/util/AsyncExecutor.scala:142]: When we re-submit a task to the thread pool, we would also need to cancel its current execution

Comment (nafg) [reply slick/src/main/scala/slick/util/AsyncExecutor.scala:142]: Good point ;)

Comment (nafg) [reply slick/src/main/scala/slick/util/AsyncExecutor.scala:142]: I guess it should rethrow then? Not sure how else to prevent its execution.

Comment (dwickern) [reply slick/src/main/scala/slick/util/AsyncExecutor.scala:142]: I don't understand this problem well enough... maybe we should have a chat on discord or something

Comment (nafg) [reply slick/src/main/scala/slick/util/AsyncExecutor.scala:142]: Absolutely, can you ping me there?

File: slick/src/main/scala/slick/util/AsyncExecutor.scala [hunk 2]
```diff
@@ -62,141 +76,201 @@ object AsyncExecutor extends Logging {
...
+                  if (managedArrayBlockingQueue.attemptPrepare(this)) {
+                    try _run(new PrioritizedRunnable.SetConnectionReleased(() => connectionReleased = true))
+                    finally
+                    // If the runnable/task has released the Jdbc connection we decrease the counter again
+                      if (!managedArrayBlockingQueue.attemptCleanUp(this))
+                        logger.warn("After executing a task, the in-use count was already 0. This should not happen.")
+                  } else {
+                    logger.warn("Could not increase in-use count. Will resubmit...")
+                    executor.execute(this)
+                  }
```
Comment (dwickern) [inline slick/src/main/scala/slick/util/AsyncExecutor.scala:199]: We also need to `_run` it when `priority() == WithConnection`

Comment (nafg) [reply slick/src/main/scala/slick/util/AsyncExecutor.scala]: Silly me. Good catch, fixed

**Comment by nafg** [comment]
Thanks so much to @jgulotta for finding the bug in this PR, looks like tests are all green now!

@dwickern can you test and review again? If you're happy then it's ready to merge!

**Comment by y-yu** [comment]
Sorry, I don't reproduce the error about https://github.com/slick/slick/discussions/2432 🙏. Thanks so much for your contribution 👍

**Comment by nafg** [comment]
@y-yu meaning it seems this fixes it? Or meaning you no longer get it at all?

**Pull request merged**